lib/ui/text.dart

```diff
@@ -1871,18 +1871,24 @@ class Paragraph extends NativeFieldWrapperClass2 {
   /// on both sides. In such cases, this method will return [offset, offset+1].
   /// Word boundaries are defined more precisely in Unicode Standard Annex #29
   /// http://www.unicode.org/reports/tr29/#Word_Boundaries
-  List<int> getWordBoundary(dynamic position) {
-    // TODO(gspencergoog): have this take only a TextPosition once the framework
-    // code is calling it with that.
-    if (position is TextPosition) {
-      return _getWordBoundary(position.offset);
-    } else {
-      final int offset = position;
-      return _getWordBoundary(offset);
-    }
+  TextRange getWordBoundary(TextPosition position) {
+    final List<int> boundary = _getWordBoundary(position.offset);
+    return TextRange(start: boundary[0], end: boundary[1]);
   }
   List<int> _getWordBoundary(int offset) native 'Paragraph_getWordBoundary';
 
+  /// Returns the [start, end] of the line at the given offset.
+  ///
+  /// The newline (if any) is returned as part of the range.
+  ///
+  /// This can potentially be expensive, since it needs to compute the line
+  /// metrics, so use it sparingly.
+  TextRange getLineBoundary(TextPosition position) {
+    final List<int> boundary = _getLineBoundary(position.offset);
+    return TextRange(start: boundary[0], end: boundary[1]);
+  }
+  List<int> _getLineBoundary(int offset) native 'Paragraph_getLineBoundary';
```

Comment:
Is there a reason this is implemented in C++ and not in Dart directly? (Implementing it in Dart would reduce unnecessary work since web can also use the same implementation).

Reply:
I don't have the history on why that is, but I assume the reason is that we didn't want to have to implement, test, and maintain the code, and it already existed in the text library.  If that's the case, then I suspect that with the advent of web, it makes sense to make the investment and implement, test, and maintain it (since we'll need to do that anyhow for web).

It's possible that it is somehow linked to the text processing library that we use, though, so I'll defer to the engine team on that.

cc/ @jason-simmons @GaryQian

Reply:
In this case, it would be far more efficient to compute it in C++ instead of serializing and passing all the LineMetrics to dart, decoding, and processing it there.

This also keeps implementations consistent throughout the Paragraph API. Methods such as GetRectsForRange requires a significant computation that can only be done at engine level, and splitting implementation between dart and C++ would make it harder to maintain.